numpy/core/src/umath/dispatching.c

```diff
@@ -0,0 +1,697 @@
+/*
+ * This file implements universal function dispatching and promotion (which
+ * is necessary to happen before dispatching).
+ * This is part of the UFunc object.  Promotion and dispatching uses the
+ * following things:
+ *
+ * - operand_DTypes:  The datatypes as passed in by the user.
+ * - signature: The DTypes fixed by the user with `dtype=` or `signature=`.
+ * - ufunc._loops: A list of all ArrayMethods and promoters, it contains
+ *   tuples `(dtypes, ArrayMethod)` or `(dtypes, promoter)`.
+ * - ufunc._dispatch_cache: A cache to store previous promotion and/or
+ *   dispatching results.
```

Comment:
Should this have an upper bound?

Reply:
> Should this have an upper bound?

Yes, although it is probably hard to run into an issue.  I will mark this as resolved though, moving to the issue gh-19544, not sure how crucial it is, but attacking it in this PR is not going to help I think.